### ISSUE TITLE
Download button in playlist item does not hide properly after download finishes

### ISSUE THREAD

**Issue by hlysine** [ppy/osu#13972]
**Describe the bug:** 
1. Download multiple beatmaps in a playlist by clicking the download button in playlist items.
2. Wait for them to finish.
3. Some playlist items will end up in a bugged state where the download button is no longer visible but the "view results" button does not reflow.

This can only be reset by exiting and re-entering the playlist screen.

(The beatmap title is also overflowing, but it should be a separate issue)

**Screenshots or videos showing encountered issue:** 
![image](https://user-images.githubusercontent.com/25472513/126582466-e1f51076-1b19-43b1-bc52-9a7bef2175a9.png)

**osu!lazer version:** 2021.720.0

**Logs:**

[database.log](https://github.com/ppy/osu/files/6859675/database.log)
[network.log](https://github.com/ppy/osu/files/6859676/network.log)
[performance.log](https://github.com/ppy/osu/files/6859677/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/6859678/performance-audio.log)
[performance-draw.log](https://github.com/ppy/osu/files/6859679/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/6859680/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/6859681/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/6859682/runtime.log)
[updater.log](https://github.com/ppy/osu/files/6859683/updater.log)

**Comment by peppy** [comment]
I'm not sure if I just have good luck, but this no longer happens after #13978. It looks like a classic invalidation bug, which may have been mitigated by the addition of the `GridContainer`.

**Comment by Henry-YSLin** [comment]
It's rare but it's still happening for me on master (e8aaf4df161b3588e37bf8209e19bd52a76ff6cf) + pulling #13979 locally so the playlist screen doesn't crash
![image](https://user-images.githubusercontent.com/25472513/126612385-0d5a49cc-685e-4232-b04d-df502b608420.png)

**Comment by peppy** [comment]
Either way, I'm 99% sure this is going to be the same issue with flow invalidation failures. Not sure if we need a tracking issue for it. @smoogipoo can you confirm?

**Comment by smoogipoo** [comment]
As far as I understand the issue, this is a persistent effect. Flow invalidation issues are 1 or 2 frame things.

Are you sure the button isn’t just present and causing flow, but with alpha = 0? It overrides IsPresent after all.

**Comment by peppy** [comment]
Hmm, not sure then. I haven't been able to repro this after more than 100 tries. The `IsPresent` logic could be part of it, but that's pretty worrying since we use that same override in many other places.

I did manage to get a button stuck in "importing" state even though it did complete importing, though. May be related.

**Comment by smoogipoo** [comment]
As far as invalidations are concerned, I’m not aware of a case in which they’re incorrect. The only issue I’m aware of is single-frame issues that always return to the expected layout.

So either way this needs more investigation.

**Comment by Henry-YSLin** [comment]
It is pretty easy to repro on my side if I just go ham on a long playlist and download everything until I hit TooManyRequests.

Upon inspection, the download button has `Alpha=0` and `IsPresent=False` in the bugged state. Would that fall into the flow invalidation cases then?
![image](https://user-images.githubusercontent.com/25472513/126731955-d289554e-9af3-44fe-8186-49b77e553199.png)

**Comment by peppy** [comment]
If you can also reproduce the issue in this area here (will be less obvious since there's only one button, but may cause truncation to look incorrect?) then you can confirm whether it's invalidation related or not by resizing the window to be very tall then very wide. If it resolves during the resize then it's quite possibly invalidation related.

![20210723 125706 (dotnet)](https://user-images.githubusercontent.com/191335/126735763-7a2f18ce-8b00-4436-bbd0-e5271a76b5b9.png)

**Comment by smoogipoo** [comment]
I think I can see what’s going on - because it’s overriding, IsPresent may change without causing an invalidation.

**Comment by Henry-YSLin** [comment]
> If you can also reproduce the issue in this area here (will be less obvious since there's only one button, but may cause truncation to look incorrect?) then you can confirm whether it's invalidation related or not by resizing the window to be very tall then very wide. If it resolves during the resize then it's quite possibly invalidation related.

Weird, I can't seem to repro it in that area. I've checked with the draw visualiser that all download buttons are hidden properly.
